onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -13,7 +15,209 @@
+constexpr auto TENSOR_UINT8 = "tensor(uint8)";
+constexpr auto TENSOR_INT16 = "tensor(int16)";
+constexpr auto TENSOR_UINT16 = "tensor(uint16)";
+constexpr auto TENSOR_INT32 = "tensor(int32)";
+constexpr auto TENSOR_UINT32 = "tensor(int32)";
+constexpr auto TENSOR_INT64 = "tensor(int64)";
+constexpr auto TENSOR_UINT64 = "tensor(int64)";
+constexpr auto TENSOR_FLOAT = "tensor(float)";
+constexpr auto TENSOR_FLOAT16 = "tensor(float16)";
+constexpr auto TENSOR_DOUBLE = "tensor(double)";
+constexpr auto TENSOR_STRING = "tensor(string)";
+constexpr auto TENSOR_BOOL = "tensor(bool)";
+constexpr auto TENSOR_COMPLEX64 = "tensor(complex64)";
+constexpr auto TENSOR_COMPLEX128 = "tensor(complex128)";
+
 namespace ONNX_NAMESPACE {
+
+template<typename U>
+struct span {
+    class iterator {
+    public:
+        iterator(const U *ptr): ptr(ptr) {}
+        iterator operator++() { ++ptr; return *this; }
+        bool operator!=(const iterator& other) const { return ptr != other.ptr; }
+        const U& operator*() const { return *ptr; }
+    private:
+        U* ptr;
+    };
+
+    const U *data;
+    size_t n;
+
+    constexpr span() : data(), n() {}
+
+    template <size_t N>
+    constexpr span(const U (&arr)[N])
+    : data(arr), n(N) {}
+
+    iterator begin() const { return iterator(data); }
+    iterator end() const { return iterator(data + n); }
+    size_t size() const { return n; }
+    const U& operator[](size_t i) const { return data[i]; }
+};
+
+struct FormalParameter {
+    constexpr FormalParameter(
+            const char* name,
+            const char*,
+            const char* type_str,
+            OpSchema::FormalParameterOption param_option = OpSchema::FormalParameterOption::Single,
+            bool is_homogeneous = true,
+            int min_arity = 1) :
+            name(name),
+            type_str(type_str),
+            param_option(param_option),
+            is_homogeneous(is_homogeneous),
+            min_arity(min_arity)
+    {}
+
+    const char* name;
+    const char* type_str;
+    OpSchema::FormalParameterOption param_option;
+    bool is_homogeneous;
+    int min_arity;
+};
+
+struct Attribute {
+    constexpr Attribute(
+            const char* name,
+            const char*,
+            AttributeProto::AttributeType type,
+            bool required = true)
+            : name_(name),
+              type_(type),
+              required_(required) {}
+
+    constexpr Attribute(
+            const char* name,
+            const char*,
+            AttributeProto::AttributeType type,
+            float /*default_value*/)
+            : name_(name),
+              type_(type) {
+        // TODO populate default_value
+    }
+
+    constexpr Attribute(
+            const char* name,
+            const char*,
+            AttributeProto::AttributeType type,
+            int64_t /*default_value*/)
+            : name_(name),
+              type_(type) {
+        // TODO populate default_value
+    }
```

Comment:
I left these two as TODOs, but they should be easy (we can hold a union of possible default values, and then create the correct AttributeProto from it when making the OpSchema)